vmm/src/lib.rs

```diff
@@ -990,181 +990,184 @@ impl Vmm {
-                }
-                info!("Migration complete");
-                Ok(())
-            } {
-                // Stop logging dirty pages and shutdown the source VM paused unpon successful migration
-                Ok(()) => {
-                    // Let every Migratable object know about the migration being complete
-                    vm.migration_complete()?;
+        // Try at most 5 passes of dirty memory sending
+        const MAX_DIRTY_MIGRATIONS: usize = 5;
+        for i in 0..MAX_DIRTY_MIGRATIONS {
+            info!("Dirty memory migration {} of {}", i, MAX_DIRTY_MIGRATIONS);
+            if !Self::vm_maybe_send_dirty_pages(vm, &mut socket)? {
+                break;
+            }
+        }
 
-                    // Stop logging dirty pages
-                    vm.stop_dirty_log()?;
+        // Now pause VM
+        vm.pause()?;
 
-                    // Shutdown the VM after the migration succeeded
-                    self.exit_evt.write(1).map_err(|e| {
-                        MigratableError::MigrateSend(anyhow!(
-                            "Failed shutting down the VM after migration: {:?}",
-                            e
-                        ))
-                    })?;
+        // Send last batch of dirty pages
+        Self::vm_maybe_send_dirty_pages(vm, &mut socket)?;
 
-                    Ok(())
-                }
-                // Ensure the source VM continue to run upon unsuccessful migration
-                Err(e) => {
-                    error!("Migration failed: {:?}", e);
+        // Capture snapshot and send it
+        let vm_snapshot = vm.snapshot()?;
+        let snapshot_data = serde_json::to_vec(&vm_snapshot).unwrap();
+        Request::state(snapshot_data.len() as u64).write_to(&mut socket)?;
+        socket
+            .write_all(&snapshot_data)
+            .map_err(MigratableError::MigrateSocket)?;
+        let res = Response::read_from(&mut socket)?;
+        if res.status() != Status::Ok {
+            warn!("Error during state migration");
+            Request::abandon().write_to(&mut socket)?;
+            Response::read_from(&mut socket).ok();
+            return Err(MigratableError::MigrateSend(anyhow!(
+                "Error during state migration"
+            )));
+        }
+
+        // Complete the migration
+        Request::complete().write_to(&mut socket)?;
+        let res = Response::read_from(&mut socket)?;
+        if res.status() != Status::Ok {
+            warn!("Error completing migration");
+            Request::abandon().write_to(&mut socket)?;
+            Response::read_from(&mut socket).ok();
+            return Err(MigratableError::MigrateSend(anyhow!(
+                "Error completing migration"
+            )));
+        }
+        info!("Migration complete");
+
+        // Let every Migratable object know about the migration being complete
+        vm.migration_complete()?;
+
+        // Stop logging dirty pages
+        vm.stop_dirty_log()?;
```

Comment:
It seems to be redundant, if we decide to go with shutting-down the source VM upon successful migration (also we explicitly stop logging dirty pages upon failed migration below).

Reply:
Let's keep it in case we decide to change the behavior about shutting down the VM in the future. And this provides a clean uninitialization path.